test/distributed/org/apache/cassandra/distributed/test/PagingTest.java

```diff
@@ -0,0 +1,77 @@
+            singleNode.schemaChange(withKeyspace("CREATE TABLE %s.tbl (pk int, ck int, v int, PRIMARY KEY (pk, ck))"));
+
+            for (int i = 0; i < 10; i++)
+            {
+                for (int j = 0; j < 10; j++)
+                {
+                    cluster.coordinator(1)
+                           .execute(withKeyspace("INSERT INTO %s.tbl (pk, ck, v) VALUES (1, ?, ?)"), QUORUM, i, j, i + i);
+                    singleNode.coordinator(1)
+                              .execute(withKeyspace("INSERT INTO %s.tbl (pk, ck, v) VALUES (1, ?, ?)"), QUORUM, i, j, i + i);
+                }
+            }
+
+            int[] pageSizes = new int[]{ 1, 2, 3, 5, 10, 20, 50 };
+            String[] statements = new String[]{ withKeyspace("SELECT * FROM %s.tbl WHERE pk = 1 AND ck > 5"),
+                                                withKeyspace("SELECT * FROM %s.tbl WHERE pk = 1 AND ck >= 5"),
+                                                withKeyspace("SELECT * FROM %s.tbl WHERE pk = 1 AND ck > 5 AND ck <= 10"),
+                                                withKeyspace("SELECT * FROM %s.tbl WHERE pk = 1 AND ck > 5 LIMIT 3"),
+                                                withKeyspace("SELECT * FROM %s.tbl WHERE pk = 1 AND ck >= 5 LIMIT 2"),
+                                                withKeyspace("SELECT * FROM %s.tbl WHERE pk = 1 AND ck > 5 AND ck <= 10 LIMIT 2"),
+                                                withKeyspace("SELECT * FROM %s.tbl WHERE pk = 1 AND ck > 5 ORDER BY ck DESC"),
+                                                withKeyspace("SELECT * FROM %s.tbl WHERE pk = 1 AND ck >= 5 ORDER BY ck DESC"),
+                                                withKeyspace("SELECT * FROM %s.tbl WHERE pk = 1 AND ck > 5 AND ck <= 10 ORDER BY ck DESC"),
+                                                withKeyspace("SELECT * FROM %s.tbl WHERE pk = 1 AND ck > 5 ORDER BY ck DESC LIMIT 3"),
+                                                withKeyspace("SELECT * FROM %s.tbl WHERE pk = 1 AND ck >= 5 ORDER BY ck DESC LIMIT 2"),
+                                                withKeyspace("SELECT * FROM %s.tbl WHERE pk = 1 AND ck > 5 AND ck <= 10 ORDER BY ck DESC LIMIT 2"),
+                                                withKeyspace("SELECT DISTINCT pk FROM %s.tbl LIMIT 3"),
+                                                withKeyspace("SELECT DISTINCT pk FROM %s.tbl WHERE pk IN (3,5,8,10)"),
+                                                withKeyspace("SELECT DISTINCT pk FROM %s.tbl WHERE pk IN (3,5,8,10) LIMIT 2")
+            };
+            for (String statement : statements)
+            {
+                for (int pageSize : pageSizes)
+                {
+                    assertRows(cluster.coordinator(1).executeWithPaging(statement, QUORUM, pageSize),
+                               singleNode.coordinator(1).executeWithPaging(statement, QUORUM, Integer.MAX_VALUE));
```

Comment:
There's no interplay between SELECT, paging, and range tombstones (plus reversed data types), but I think it's best we leave it to automated/fuzz tests.

Reply:
> non-paged response will be same across all page sizes.

Good catch. This test has been moved with minimal changes from [`SimpleReadWriteTest#pagingTests`](https://github.com/apache/cassandra/blob/trunk/test/distributed/org/apache/cassandra/distributed/test/SimpleReadWriteTest.java#L276-L329) during a final refactor to give `SimpleReadWriteTest` the responsibilities of the Python dtest `TestPutGet`. I guess that it was done this way to use the version of `assertRows` that accepts row iterators, that would exhaust the iterator for the non-paged query if we reuse it. I have changed it to use the version of `assertRows` that accepts arrays so we can reuse the non-paged results.

> There's no interplay between SELECT, paging, and range tombstones (plus reversed data types), but I think it's best we leave it to automated/fuzz tests.

`ShortReadProtectionTest`, `ReadRepairQueryTypesTest` and `ReadRepairEmptyRangeTombstonesTest` include some interaction with paging and reversed types (that's how we found CASSANDRA-15459), but it's true that it would be great to have that more systematically covered with fuzz tests.